### PR TITLE
Avoid possible truncation of higher bits

### PR DESCRIPTION
**Pull request by xgreenx**
Closes https://github.com/FuelLabs/fuel-vm/issues/570

### CODE REVIEW

File: fuel-tx/src/transaction/types/output.rs [hunk 1]
```diff
@@ -181,7 +181,11 @@ impl Output {
     }
 
     pub fn message_nonce(txid: &Bytes32, idx: Word) -> Nonce {
-        (*Hasher::default().chain(txid).chain([idx as u8]).finalize()).into()
+        (*Hasher::default()
+            .chain(txid)
+            .chain(idx.to_bytes())
```
Comment (xgreenx) [inline fuel-tx/src/transaction/types/output.rs:186]: If we are okay with supporting more than `255` receipts, then we also need to update the [specification](https://github.com/FuelLabs/fuel-specs/blob/master/src/identifiers/utxo-id.md#message-nonce)

File: fuel-merkle/src/common/msb.rs [hunk 2]
```diff
@@ -62,11 +62,10 @@ mod test {
         Bytes8,
         Msb,
     };
-    use core::mem::size_of;
 
     #[test]
     fn test_msb_for_bytes_1() {
-        const NUM_BITS: usize = size_of::<Bytes1>() * 8;
```
Comment (Voxelot) [inline fuel-merkle/src/common/msb.rs:69]: I like the `size_of` approach as it's more obvious to the code reader how this value is constructed. Maybe we should split  this into two constants like `NUM_BYTES` vs `NUM_BITS`?

File: fuel-tx/src/transaction/types/output.rs (hunk 1 above)
Comment (Dentosal) [reply fuel-tx/src/transaction/types/output.rs:186]: I think this was settled yesterday and we do want to support U16::MAX receipts. Would a PR changing the linked paragraph be enough? I recall there was some discussion about serialization as well.

Comment (xgreenx) [reply fuel-tx/src/transaction/types/output.rs:186]: We need to update the paragraph and mention that we need to use canonical serialization for integers. 

We have a separate issue to add a `u16::max` limit https://github.com/FuelLabs/fuel-vm/issues/620

**Review by Dentosal — APPROVED** [review]

File: fuel-merkle/src/common/msb.rs [hunk 3]
```diff
@@ -5,11 +5,11 @@ pub enum Bit {
 }
 
 trait GetBit {
-    fn get_bit(&self, bit_index: usize) -> Option<Bit>;
+    fn get_bit(&self, bit_index: u32) -> Option<Bit>;
```
Comment (bvrooman) [inline fuel-merkle/src/common/msb.rs:8]: Are these changes to `fuel-merkle` necessary? It feels like we're changing library logic to fit business logic. A library should promise correctness under all targets it supports, including all 32 bit and 64 bit targets. If it has internal problems with truncation, that should be fixed in the library. But this change appears to be changing the library interface to accommodate changes to external libraries.

Comment (xgreenx) [reply fuel-merkle/src/common/msb.rs:8]: It is a private trait, so it is an internal detail of how Merkle tree works=)

**Pull request merged**